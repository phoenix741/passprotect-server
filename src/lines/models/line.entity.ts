import { Document } from 'mongoose';
import { EncryptedContent } from '../../shared/interfaces/encrypted-content.interface';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';
import { ObjectID } from 'bson';

export interface LineEntity extends Document {
  _id: ObjectID;
  type: LineTypeEnum;
  label: string;
  group: string;
  logo?: Buffer;
  encryption: EncryptedContent;
  user: string;
  createdAt: Date;
  updatedAt: Date;
  _rev: number;
}
