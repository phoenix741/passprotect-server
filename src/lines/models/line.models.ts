import { EncryptedContent } from '../../shared/interfaces/encrypted-content.interface';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';
import { ObjectID } from 'bson';

export interface LineToCreate {
  label: string;
  type: LineTypeEnum;
  group: string;
  encryption: EncryptedContent;
  user: string;
}

export interface LineToUpdate {
  _id?: ObjectID;
  label: string;
  type: LineTypeEnum;
  group: string;
  encryption: EncryptedContent;
}
