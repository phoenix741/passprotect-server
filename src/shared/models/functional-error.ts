import { HttpException } from '@nestjs/common';

export class FunctionalError extends Error {
  field?: string;

  constructor(field?: string, message?: string) {
    super(message);

    this.field = field;
  }

  toGraphQL(defaultLabel = 'label') {
    return {
      errors: [{
        fieldName: this.field || defaultLabel,
        message: this.message,
      }],
    };
  }
}

export function toFunctionalError(error: Error): FunctionalError {
  if (error instanceof FunctionalError) {
    return error;
  }
  if (error instanceof HttpException) {
    return new FunctionalError(undefined, error.message.message);
  }
  return new FunctionalError(undefined, error.message);
}
