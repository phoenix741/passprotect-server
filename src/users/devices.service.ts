import { Model } from 'mongoose';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { UserEntity } from './models/user.entity';
import { DeviceEntity } from './models/user.entity';
import { UsersService } from './users.service';

@Injectable()
export class DevicesService {
  constructor(private readonly userService: UsersService) {}

  async findAll(userId: string): Promise<DeviceEntity[]> {
    if (userId) {
      const user = await this.userService.findById(userId);
      return user.devices;
    }
    return [];
  }

  async verifyDevice(user: UserEntity, deviceId: string): Promise<UserEntity> {
    const isValid = user.devices.find(d => d.id === deviceId);
    if (!isValid) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }
    return user;
  }

  async registerDevice(userId: string, device: DeviceEntity): Promise<DeviceEntity> {
    const user = await this.userService.findById(userId);
    if (!user) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }

    const found = user.devices.find(d => d.id === device.id);
    if (!found) {
      user.devices.push(device);
      await user.save();
    }
    return found || device;
  }

  async removeDevice(userId: string, deviceId: string): Promise<void> {
    const user = await this.userService.findById(userId);
    if (!user) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }

    user.devices = user.devices.filter(d => d.id !== deviceId);
    await user.save();
  }
}
