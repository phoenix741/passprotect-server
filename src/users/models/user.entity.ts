import { Document } from 'mongoose';
import { EncryptedContent } from '../../shared/interfaces/encrypted-content.interface';

export interface UserEntity extends Document {
  _id: string;
  password: string;
  encryption: EncryptedContent;
  createdAt: Date;
  updatedAt: Date;
}
