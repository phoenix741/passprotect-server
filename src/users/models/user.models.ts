import { EncryptedContent } from '../../shared/interfaces/encrypted-content.interface';

export interface UserToRegister {
  _id: string;
  password: string;
  encryption: EncryptedContent;
}

export interface JwtPayload {
  sub: string;
  issuer: string;
}
