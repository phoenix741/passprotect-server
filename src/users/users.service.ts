import { Model } from 'mongoose';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { UserToRegister } from './models/user.models';
import { UserEntity } from './models/user.entity';
import { CryptoService } from '../shared/services/crypto.service';
import { DeviceEntity } from '../users/models/user.entity';

@Injectable()
export class UsersService {
  constructor(private readonly cryptoService: CryptoService, @InjectModel('User') private readonly userModel: Model<UserEntity>) {}

  async findAll(): Promise<UserEntity[]> {
    return await this.userModel.find().exec();
  }

  async findById(id: string): Promise<UserEntity | null> {
    return id && (await this.userModel.findById(id.toLowerCase()));
  }

  async registerUser(user: UserToRegister): Promise<UserEntity> {
    user.password = await this.cryptoService.hashPassword(user.password);

    return await this.userModel.create(user);
  }

  async verifyPassword(user: UserEntity, password: string): Promise<UserEntity> {
    const isValid = await this.cryptoService.checkPassword(password, user.password);
    if (!isValid) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }
    return user;
  }
}
