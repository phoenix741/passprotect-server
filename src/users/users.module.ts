import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { UserSchema } from './schemas/user.schema';
import { UsersService } from './users.service';
import { SharedModule } from '../shared/shared.module';
import { DevicesService } from './devices.service';

@Module({
  imports: [MongooseModule.forFeature([{ name: 'User', schema: UserSchema }]), SharedModule],
  providers: [UsersService, DevicesService],
  exports: [UsersService, DevicesService],
})
export class UsersModule {}
