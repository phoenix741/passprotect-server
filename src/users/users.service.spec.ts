import { Test } from '@nestjs/testing';
import * as mongoose from 'mongoose';
import { Model } from 'mongoose';
import mockingoose from 'mockingoose';
import { UserSchema } from './schemas/user.schema';
import { UserEntity } from './models/user.entity';
import { getModelToken } from '@nestjs/mongoose';
import { UsersService } from './users.service';
import { CryptoService } from '../shared/services/crypto.service';
import { UserToRegister } from './models/user.models';

describe('UsersService', () => {
  let userModel: Model<UserEntity>;
  let usersService: UsersService;
  let cryptoService;

  const doc = {
    _id: 'username',
    password: 'password',
  };

  beforeEach(async () => {
    userModel = mongoose.model('User', UserSchema);
    cryptoService = {
      hashPassword: jest
        .fn()
        .mockImplementation(password => `hashed${password}`),
      checkPassword: jest.fn().mockImplementation((p1, p2) => p1 === p2),
    };

    const module = await Test.createTestingModule({
      providers: [
        UsersService,
        {
          provide: CryptoService,
          useValue: cryptoService,
        },
        {
          provide: getModelToken('User'),
          useValue: userModel,
        },
      ],
    }).compile();

    usersService = module.get<UsersService>(UsersService);
  });

  describe('findAll', () => {
    it('success', async () => {
      mockingoose(userModel).toReturn([doc], 'find');
      const users = await usersService.findAll();
      expect(users.map(user => user.toObject())).toMatchSnapshot(
        'result of find all users',
      );
    });
  });

  describe('findById', () => {
    it('success, with id', async () => {
      mockingoose(userModel).toReturn(doc, 'findOne');
      const user = await usersService.findById('username');
      expect(user.toObject()).toMatchSnapshot('result of find user by id');
    });

    it('success, without id', async () => {
      mockingoose(userModel).toReturn(doc, 'findOne');
      const user = await usersService.findById(null);
      expect(user).toBeNull();
    });
  });

  describe('registerUser', () => {
    it('success', async () => {
      const user: UserToRegister = {
        _id: 'username',
        password: 'password',
        encryption: {
          salt: Buffer.from('salt'),
          iv: Buffer.from('iv'),
          authTag: Buffer.from('authTag'),
          content: Buffer.from('content'),
        },
      };
      mockingoose(userModel).toReturn((userToCreate: UserEntity) => {
        expect(userToCreate.toObject()).toMatchSnapshot(
          { createdAt: expect.any(Date), updatedAt: expect.any(Date) },
          'user to create',
        );
        return user;
      }, 'save');
      const userCreated = await usersService.registerUser(user);
      expect(userCreated).toMatchSnapshot('registered user');
    });
  });

  describe('verifyPassword', () => {
    it('success', async () => {
      const user: UserEntity = {
        _id: 'username',
        password: 'password',
      } as UserEntity;
      const isValid = await usersService.verifyPassword(user, 'password');
      expect(isValid).toEqual(user);
    });

    it('failed', async () => {
      const user: UserEntity = {
        _id: 'username',
        password: 'password',
      } as UserEntity;
      await expect(
        usersService.verifyPassword(user, 'failed'),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
