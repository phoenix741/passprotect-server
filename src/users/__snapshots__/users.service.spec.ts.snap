// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersService findAll success: result of find all users 1`] = `
Array [
  Object {
    "_id": "username",
    "password": "password",
  },
]
`;

exports[`UsersService findById success, with id: result of find user by id 1`] = `
Object {
  "_id": "username",
  "password": "password",
}
`;

exports[`UsersService registerUser success: registered user 1`] = `
Object {
  "_id": "username",
  "encryption": Object {
    "authTag": Object {
      "data": Array [
        97,
        117,
        116,
        104,
        84,
        97,
        103,
      ],
      "type": "Buffer",
    },
    "content": Object {
      "data": Array [
        99,
        111,
        110,
        116,
        101,
        110,
        116,
      ],
      "type": "Buffer",
    },
    "iv": Object {
      "data": Array [
        105,
        118,
      ],
      "type": "Buffer",
    },
    "salt": Object {
      "data": Array [
        115,
        97,
        108,
        116,
      ],
      "type": "Buffer",
    },
  },
  "password": "hashedpassword",
}
`;

exports[`UsersService registerUser success: user to create 1`] = `
Object {
  "_id": "username",
  "createdAt": Any<Date>,
  "encryption": Object {
    "authTag": "YXV0aFRhZw==",
    "content": "Y29udGVudA==",
    "iv": "aXY=",
    "salt": "c2FsdA==",
  },
  "password": "hashedpassword",
  "updatedAt": Any<Date>,
}
`;

exports[`UsersService verifyPassword failed 1`] = `
Object {
  "error": "Unauthorized",
  "message": "error:user.404.userNotFound",
  "statusCode": 401,
}
`;
